core/src/avm2/optimize.rs

```diff
@@ -0,0 +1,893 @@
+                            if !class.inner_class_definition().read().is_interface() {
+                                match class.instance_vtable().get_trait(&multiname) {
+                                    Some(Property::Slot { slot_id }) => {
+                                        *op = Op::SetSlot { index: slot_id };
+                                    }
+                                    _ => {}
+                                }
+                            }
+                        }
+                    }
+                    // `stack_pop_multiname` handled lazy
+                }
+            }
+            Op::Construct { num_args } => {
+                // Arguments
+                stack.popn(*num_args);
+
+                stack.pop();
+
+                // Avoid checking return value for now
+                stack.push_any();
+            }
+            Op::ConstructSuper { num_args } => {
+                // Arguments
+                stack.popn(*num_args);
+
+                // Then receiver.
+                stack.pop();
+            }
+            Op::ConstructProp {
+                index: name_index,
+                num_args,
+            } => {
+                // Arguments
+                stack.popn(*num_args);
+
+                stack_pop_multiname!(name_index);
+
+                // Then receiver.
+                stack.pop();
+
+                // Avoid checking return value for now
+                stack.push_any();
+            }
+            Op::CallProperty {
+                index: name_index,
+                num_args,
+            } => {
+                // Arguments
+                stack.popn(*num_args);
+
+                let multiname = stack_pop_multiname!(name_index);
+
+                // Then receiver.
+                let stack_value = stack.pop();
+
+                if let Some(multiname) = multiname {
+                    if !multiname.has_lazy_component() {
+                        if let Some(ValueType::Class(class)) = stack_value {
+                            if !class.inner_class_definition().read().is_interface() {
+                                match class.instance_vtable().get_trait(&multiname) {
+                                    Some(Property::Method { disp_id }) => {
+                                        *op = Op::CallMethod {
+                                            num_args: *num_args,
+                                            index: Index::new(disp_id),
+                                        };
+                                    }
+                                    _ => {}
+                                }
+                            }
+                        }
+                    }
+                    // `stack_pop_multiname` handled lazy
+                }
+
+                // Avoid checking return value for now
+                stack.push_any();
+            }
+            Op::CallPropVoid { .. } => {
```

Comment:
IIRC, `CallPropVoid` is identical, except it doesn't push at the end?
How about having the code above match both and just check at the end?

(AFAIK `CallPropLex` is also conceptually very similar to `CallProperty` - I think it just doesn't set `this` on the called function...?)

Reply:
There's no `CallMethodVoid`, so the optimization would also have to be disabled. I think this can wait until later- I don't think any compiler would emit a `CallPropVoid` with values still on the stack.